lib/lotus/validations/error.rb

```diff
@@ -63,21 +60,16 @@ class Error
 
       # Initialize a validation error
       #
-      # @param attribute_name [Symbol] the name of the attribute
       # @param validation [Symbol] the name of the validation
       # @param expected [Object] the expected value
       # @param actual [Object] the actual value
-      # @param namespace [String] the optional namespace
       #
       # @since 0.1.0
       # @api private
-      def initialize(attribute_name, validation, expected, actual, namespace = nil)
-        @attribute_name = attribute_name.to_s
+      def initialize(validation, expected, actual)
```

Comment:
@rodzyn Consider the following code, it uses v0.3.2

```ruby
 ~:ruby-2.2.2 ➜ pry
[1] pry(main)> require 'lotus/validations'
=> true
[2] pry(main)> class Signup
[2] pry(main)*   include Lotus::Validations
[2] pry(main)*   attribute :user do
[2] pry(main)*     attribute :name, presence: true
[2] pry(main)*     attribute :address do
[2] pry(main)*       attribute :city, presence: true
[2] pry(main)*     end
[2] pry(main)*   end
[2] pry(main)*   attribute :terms, acceptance: true
[2] pry(main)* end
=> {:acceptance=>true}
[3] pry(main)> signup = Signup.new
=> #<Signup:0x007fa4c4c4bb28 @attributes=#<Lotus::Utils::Attributes:0x007fa4c4c4bad8 @attributes={}>>
[4] pry(main)> signup.valid?
=> false
[5] pry(main)> signup.errors
=> #<Lotus::Validations::Errors:0x007fa4c4c23998
 @errors=
  {"user.name"=>
    [#<Lotus::Validations::Error:0x007fa4c4c22c50 @actual=nil, @attribute="user.name", @attribute_name="name", @expected=true, @namespace=:user, @validation=:presence>],
   "user.address.city"=>
    [#<Lotus::Validations::Error:0x007fa4c4c22ae8
      @actual=nil,
      @attribute="user.address.city",
      @attribute_name="address.city",
      @expected=true,
      @namespace=:user,
      @validation=:presence>],
   :terms=>[#<Lotus::Validations::Error:0x007fa4c4c228e0 @actual=nil, @attribute="terms", @attribute_name="terms", @expected=true, @namespace=nil, @validation=:acceptance>]}>
[6] pry(main)> signup.errors.each do |error|
[6] pry(main)*   puts error.attribute
[6] pry(main)*   puts error.validation
[6] pry(main)*   puts "\n\n\n"
[6] pry(main)* end
user.name
presence



user.address.city
presence



terms
acceptance



=> [#<Lotus::Validations::Error:0x007fa4c4c22c50 @actual=nil, @attribute="user.name", @attribute_name="name", @expected=true, @namespace=:user, @validation=:presence>,
 #<Lotus::Validations::Error:0x007fa4c4c22ae8
  @actual=nil,
  @attribute="user.address.city",
  @attribute_name="address.city",
  @expected=true,
  @namespace=:user,
  @validation=:presence>,
 #<Lotus::Validations::Error:0x007fa4c4c228e0 @actual=nil, @attribute="terms", @attribute_name="terms", @expected=true, @namespace=nil, @validation=:acceptance>]
[7] pry(main)>
```

The advantage of having flatten errors is the use only one level of `each`.

With your suggestion you need to know exactly how many levels of depth and construct nested loops. In other words, the way we should iterate for errors, will change, if we change the nested structure of the validations. Which is an annoying maintenance issue.

```ruby
irb(main):001:0> class Signup
irb(main):002:1> include Lotus::Validations
irb(main):003:1> attribute :user do
irb(main):004:2* attribute :name, presence: true
irb(main):005:2> attribute :address do
irb(main):006:3* attribute :city, presence: true
irb(main):007:3> end
irb(main):008:2> end
irb(main):009:1> attribute :terms, acceptance: true
irb(main):010:1> end
=> {:acceptance=>true}
irb(main):011:0> signup = Signup.new
=> #<Signup:0x007fee19d9cdf8 @attributes=#<Lotus::Utils::Attributes:0x007fee19d9cd80 @attributes={}>>
irb(main):012:0> signup.valid?
=> false
irb(main):013:0> signup.errors
=> #<Lotus::Validations::Errors:0x007fee19d95c38 @errors={"user.name"=>[#<Lotus::Validations::Error:0x007fee19d958a0 @validation=:presence, @expected=true, @actual=nil>], "user.address.city"=>[#<Lotus::Validations::Error:0x007fee19d955a8 @validation=:presence, @expected=true, @actual=nil>], :terms=>[#<Lotus::Validations::Error:0x007fee19d94f90 @validation=:acceptance, @expected=true, @actual=nil>]}>
irb(main):014:0> signup.errors.each do |attr_name, errors|
irb(main):015:1* puts attr_name
irb(main):016:1> errors.each do |error|
irb(main):017:2* puts error.validation
irb(main):018:2> end
irb(main):019:1> puts "\n\n\n"
irb(main):020:1> end
user.name
presence



user.address.city
presence



terms
acceptance



=> {"user.name"=>[#<Lotus::Validations::Error:0x007fee19d958a0 @validation=:presence, @expected=true, @actual=nil>], "user.address.city"=>[#<Lotus::Validations::Error:0x007fee19d955a8 @validation=:presence, @expected=true, @actual=nil>], :terms=>[#<Lotus::Validations::Error:0x007fee19d94f90 @validation=:acceptance, @expected=true, @actual=nil>]}
```

Imagine we want to introduce an `ErrorPresenter` (which is something we are considering). Its constructor should accept an `Error` and an attribute name. It looks like a Data Clump. Error is a type designed to remove this problem and have only one object to hold all the informations, without any external reference (the attr name).